### PR TITLE
feat(chat-ui): add smart apply to chat panel

### PR DESCRIPTION
**Pull request by Sma1lboy**
### Demo (last update: 10/17/2024):

https://github.com/user-attachments/assets/e8b779fe-c0c6-41c4-a149-4341acce702b

Three general way using smart apply
- adding tests
- adding comments
- insert similar code

### Steps
- Implement a smart apply button in the chat view
- Use fuzzy search to get line range for insertion or replacement
- If fuzzy search cannot find an appropriate line range, use the Tabby backend to prompt for a range to insert
- Refactor chat useful functions
- Refactor Workspace operation requests into `WorkspaceFeature` in VSCode
- Keep `provideSmartApplyEdit`, consider still using it in situations where the current smartApply is not working
- Change the prompt for provideSmartApplyLineRange for the Tabby backend

### CODE REVIEW

File: clients/tabby-chat-panel/src/index.ts [hunk 1]
```diff
@@ -50,6 +50,9 @@ export interface ClientApi {
 
   onApplyInEditor: (content: string) => void
 
+  // smart apply into editor
+  onSmartApplyInEditor: (languageId: string, content: string) => void
```
Comment (wsxiaoys) [inline clients/tabby-chat-panel/src/index.ts:54]:
```suggestion
  onApplyInEditor: (content: string, opts?: { languageId: string, smart: boolean}) => void
```

File: clients/vscode/src/lsp/WorkspaceFeature.ts [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+      this.client.onRequest(ApplyWorkspaceEditRequest.type, (params: ApplyWorkspaceEditParams) => {
+        return this.handleApplyWorkspaceEdit(params);
+      }),
+      this.client.onRequest(RevealEditorRangeRequest.type, (params: RevealEditorRangeParams) => {
+        return this.handleRevealEditorRange(params);
+      }),
+    );
+  }
+
+  register(data: RegistrationData<unknown>): void {
```
Comment (icycodes) [inline clients/vscode/src/lsp/WorkspaceFeature.ts:45]: `register` and `unregister` is not work with `StaticFeature`

**Comment by icycodes** [comment]
@Sma1lboy, could you please rebase this pull request on the main branch and resolve conflicts?

File: clients/vscode/src/lsp/WorkspaceFeature.ts [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    // nothing
+  }
+
+  initialize(): void {
+    this.disposables.push(
+      this.client.onRequest(ApplyWorkspaceEditRequest.type, (params: ApplyWorkspaceEditParams) => {
+        return this.handleApplyWorkspaceEdit(params);
+      }),
+      this.client.onRequest(ShowDocumentRequest.type, (params: ShowDocumentParams) => {
+        return this.handleRevealEditorRange(params);
```
Comment (icycodes) [inline clients/vscode/src/lsp/WorkspaceFeature.ts:43]: I thought this standard LSP method should already be implemented by VSCode, so we do not need to implement it again. Could you help confirm this?

Comment (Sma1lboy) [reply clients/vscode/src/lsp/WorkspaceFeature.ts:43]: ye sure, give me a second

**Comment by Sma1lboy** [comment]
> @Sma1lboy, could you please rebase this pull request on the main branch and resolve conflicts?

Hi @icycodes, I just rebased the code.

**Comment by icycodes** [comment]
Hi @Sma1lboy, I have created a PR at https://github.com/Sma1lboy/tabby/pull/2 that makes some updates to the code style. Please check it out. Thank you.

**Pull request merged**